core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,278 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.accumulo.core.data;
+
+
+import org.apache.hadoop.io.Text;
+
+/**
+ * A builder used to build <code>Key</code>s by defining their components.
+ *
+ * The rules are:
+ * <ul>
+ *   <li>All components of the <code>Key</code> are optional except the row</li>
+ *   <li>Components not explicitly set default to empty byte array except the timestamp which
+ *   defaults to <code>Long.MAX_VALUE</code></li>
+ *   <li>The column qualifier can only be set if the column family has been set first</li>
+ *   <li>The column visibility can only be set if at least the column family has been set first</li>
+ * </ul>
+ *
+ * The builder supports three types of components: <code>byte[]</code>, <code>Text</code> and <code>CharSequence</code>.
+ *
+ * The builder is mutable and not thread safe.
+ *
+ * @see org.apache.accumulo.core.data.Key
+ * @since 1.8
+ */
+public class KeyBuilder {
+
+  public interface Build<T> {
+
+    /**
+     * Build a <code>Key</code> from this builder.
+     *
+     * @param copyBytes
+     *          if the byte arrays should be copied or not. If not, byte arrays will be reused in the resultant <code>Key</code>
+     * @return
+     *          the <code>Key</code> built from this builder
+     */
+    Key build(boolean copyBytes);
+
+    /**
+     * Build a <code>Key</code> from this builder. copyBytes defaults to true.
+     *
+     * @return
+     *          the <code>Key</code> built from this builder
+     */
+    Key build();
+
+    /**
+     * Change the timestamp of the <code>Key</code> created.
+     *
+     * @param timestamp
+     *          the timestamp to use for the <code>Key</code>
+     * @return this builder
+     */
+    Build<T> timestamp(long timestamp);
+
+    /**
+     * Set the deleted marker of the <code>Key</code> to the parameter.
+     *
+     * @param deleted
+     *          if the <code>Key</code> should be marked as deleted or not
+     * @return this builder
+     */
+    Build<T> deleted(boolean deleted);
+  }
+
+  public interface ColumnFamilyStep<T> extends ColumnVisibilityStep<T> {
```

Comment:
The column family has to be set to something.   We could make it default to empty string.  Personally I don't think we should provide a short cut in the API for setting the family to an empty string.  For this fringe case the user can just pass an empty string.